fairscale/experimental/nn/sync_batchnorm.py

```diff
@@ -0,0 +1,170 @@
+else:
+    # Copied from https://github.com/pytorch/pytorch/blob/v1.8.1/torch/distributed/nn/functional.py
+    class _AllReduce(torch.autograd.Function):
+        @staticmethod
+        def forward(ctx, op, group, tensor):  # type: ignore
+            ctx.group = group
+            ctx.op = op
+            tensor = tensor.clone()
+            dist.all_reduce(tensor, op=op, group=group)
+            return tensor
+
+        @staticmethod
+        def backward(ctx, grad_output):  # type: ignore
+            return (None, None) + (_AllReduce.apply(ctx.op, ctx.group, grad_output),)
+
+    def all_reduce(tensor, op=dist.ReduceOp.SUM, group=dist.group.WORLD):  # type: ignore
+        return _AllReduce.apply(op, group, tensor)
+
+
+def _forward(
+    input: torch.Tensor,
+    affine: bool,
+    track_running_stats: bool,
+    mean: torch.Tensor,
+    var: torch.Tensor,
+    momentum: float,
+    eps: float,
+    weight_param: torch.Tensor,
+    bias_param: torch.Tensor,
+    running_mean: torch.Tensor,
+    running_var: torch.Tensor,
+    total_count: torch.Tensor,
+) -> torch.Tensor:
+    scale = torch.rsqrt(var + eps)
+    bias = -mean * scale
+    if affine:
+        scale = scale * weight_param.reshape(mean.shape)
+        bias = bias + bias_param.reshape(mean.shape)
+
+    if track_running_stats:
+        with torch.no_grad():
+            unbiased_var = var * (total_count / (total_count - 1))
+            running_mean += momentum * (mean.reshape(-1) - running_mean)
+            running_var += momentum * (unbiased_var.reshape(-1) - running_var)
+
+    return input * scale + bias
+
+
+if torch.__version__.split(".")[:2] >= ["1", "7"]:
+    _forward = torch.jit.script(_forward)  # type: ignore
+
+
+class SyncBatchNorm(torch.nn.BatchNorm2d):
+    """
+    Fast re-implementation of ``torch.nn.SyncBatchNorm`` that can achieve a speedup
+    of 5x or more over the default implementation depending on size of the input
+    and number of distributed world.
+    """
+
+    def __init__(
+        self, *args: Tuple[Any, ...], process_group: Optional[ProcessGroup] = None, **kwargs: Dict[str, Any]
+    ) -> None:
+        super().__init__(*args, **kwargs)  # type: ignore
+        self._process_group = process_group if process_group is not None else dist.group.WORLD
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
+        if not dist.is_initialized() or not self.training:
+            return super().forward(input)
+
+        dim = [d for d in range(input.ndim) if d != 1]
+        count = torch.full((1,), input.numel() // input.size(1), device=input.device, dtype=input.dtype)
+        total_count = count.clone()
+        handle = all_reduce(total_count, group=self._process_group, async_op=True)
+        mean = torch.mean(input, dim=dim, keepdim=True)
+        meansqr = torch.mean(input * input, dim=dim, keepdim=True)
+        vec = torch.cat([mean, meansqr])
+        handle.wait()
+        vec = vec * (count / total_count)
```

Comment:
why don't we just divide by total_count(which should. be the global batch size * (product of all dimensions)? Maybe I don't get how we calculate mean and meansqr but I am not sure I understand why we need to multiple by count here.

Reply:
(count / total_count) is this workers contribution to the mean. This is to handle one worker having a larger batch than another. You can think of it like a probability.

So if total_count is 100 and this worker has a count of 40, it will contribute 0.4 towards the aggregate mean.

Reply:
Good idea! I considered this. You need to all_reduce proportionally to the local batch_size. I could handle that by all_reduce-ing the sum instead of the mean but then the problem is that you then overflow. I could avoid the overflow by first dividing by a "scale factor" that is order of total_count but was worried about the extra complexity that would add.